src/java.desktop/macosx/native/libawt_lwawt/awt/AWTWindow.m

```diff
@@ -631,6 +633,11 @@ - (BOOL) worksWhenModal {
 - (void) _deliverMoveResizeEvent {
 AWT_ASSERT_APPKIT_THREAD;
 
+    JNIEnv *pEnv = nil;
+    jint ret = (*jvm)->GetEnv(jvm, &pEnv, JNI_VERSION_1_4);
+    if (ret == JNI_EDETACHED) {
+        return;
+    }
```

Comment:
This seems like a completely arbitrary location in the AWT code for a test of something that would seem completely unexpected.
Can you PROVE that this is the only place that this will ever be needed ?
And if you can't, this isn't the right fix.
If you can it needs some comments.

Also : does AWT detach the AppKit thread from the VM on shutdown ?
IF not why does FX have to do it ? If it does, then could FX also crash if FX is embedded in an AWT app.

Reply:
I'm also not sure that this is the right fix. It seems like it will fix one specific place where AWT could do an upcall on the AppKit thread after FX has detached without addressing the problem as a whole.

Another possible fix would be to have AWT not use the cached JNI Env when it is not the NSApplication owner. I don't know how viable this fix would be, or whether it is the right fix, but it would prevent the crash in all possible places.

> Also : does AWT detach the AppKit thread from the VM on shutdown ?

I doubt it, since there isn't AFAIK an explicit toolkit shutdown in AWT like there is in FX.

> IF not why does FX have to do it ? If it does, then could FX also crash if FX is embedded in an AWT app.

FX only detaches the AppKit thread when it is the NSApplication owner, and thus attached the thread in the first place. In the case of FX embedded in an AWT app, it won't detach the thread.

We could explore what it would take for FX to not detach the AppKit thread from Java. At a minimum, it would require attaching the AppKit thread as a daemon (which AWT already does), and then creating a non-daemon Java thread that we would then terminate at FX toolkit shutdown. This would need to be tested carefully, but might be the cleanest solution.

One thing to note is that anything we do here will be about avoiding a crash in AWT, not about having the UI remain functional. In this case FX is the "owner" of the native Window system event loop (and NSApplication on macOS). Applications should not shutdown the FX toolkit (meaning, not call `Platform.exit`) and expect AWT to continue to work.

Reply:
> I'm also not sure that this is the right fix. It seems like it will fix one specific place where AWT could do an upcall on the AppKit thread after FX has detached without addressing the problem as a whole.
> 
> Another possible fix would be to have AWT not use the cached JNI Env when it is not the NSApplication owner. I don't know how viable this fix would be, or whether it is the right fix, but it would prevent the crash in all possible places.

That sounds like lots of places would need to be updated to do this.
> 
> > Also : does AWT detach the AppKit thread from the VM on shutdown ?
> 
> I doubt it, since there isn't AFAIK an explicit toolkit shutdown in AWT like there is in FX.
> 
> > IF not why does FX have to do it ? If it does, then could FX also crash if FX is embedded in an AWT app.
> 
> FX only detaches the AppKit thread when it is the NSApplication owner, and thus attached the thread in the first place. In the case of FX embedded in an AWT app, it won't detach the thread.

I didn't expect it to, but I was trying to understand why an AWT exit doesn't cause FX to crash. System.exit() is the AWT equivalent and AWT *could* have a shutdown hook to do the same. It just happens not to.

> 
> We could explore what it would take for FX to not detach the AppKit thread from Java. At a minimum, it would require attaching the AppKit thread as a daemon (which AWT already does), and then creating a non-daemon Java thread that we would then terminate at FX toolkit shutdown. This would need to be tested carefully, but might be the cleanest solution.

I would like to see that explored. What does SWT do ? 

> 
> One thing to note is that anything we do here will be about avoiding a crash in AWT, not about having the UI remain functional. In this case FX is the "owner" of the native Window system event loop (and NSApplication on macOS). Applications should not shutdown the FX toolkit (meaning, not call `Platform.exit`) and expect AWT to continue to work.

I guess Platform.exit for FX embedded inside AWT is a no-op ?

Reply:
We probably should replace the usage of getJNIEnv.
```
- JNIEnv *env = [ThreadUtilities getJNIEnv];
+ JNIEnv *env = [ThreadUtilities getJNIEnvUncached];
```

The getJNIEnv seems unsafe if some 3p libs like FX can detach the thread.

Reply:
> > Another possible fix would be to have AWT not use the cached JNI Env when it is not the NSApplication owner. I don't know how viable this fix would be, or whether it is the right fix, but it would prevent the crash in all possible places.
> 
> That sounds like lots of places would need to be updated to do this.

No, only in one place. The idea I had -- and did a quick prototype of -- was to change the implementation of `[ThreadUtilities getJNIEnv]` to check a flag to see whether or not AWT was the owner of the NSApplication and, if not, effectively do the same thing as `getJNIEnvUncached`. What that would then do would be to re-attach the AppKit thread to the JVM if FX had detached it. This will avoid the crash in all places without having to change the callers of `getJNIEnv`

 > We could explore what it would take for FX to not detach the AppKit thread from Java. At a minimum, it would require attaching the AppKit thread as a daemon (which AWT already does), and then creating a non-daemon Java thread that we would then terminate at FX toolkit shutdown. This would need to be tested carefully, but might be the cleanest solution.
> 
> I would like to see that explored. What does SWT do ?

I'll explore this for FX. I think it will be pretty easy. No idea what SWT does.

> > One thing to note is that anything we do here will be about avoiding a crash in AWT, not about having the UI remain functional. In this case FX is the "owner" of the native Window system event loop (and NSApplication on macOS). Applications should not shutdown the FX toolkit (meaning, not call `Platform.exit`) and expect AWT to continue to work.
> 
> I guess Platform.exit for FX embedded inside AWT is a no-op ?

Not a no-op, since it still shuts down the FX toolkit and renderer, but it doesn't detach the AppkitThread from the JVM.

Reply:
> > > Another possible fix would be to have AWT not use the cached JNI Env when it is not the NSApplication owner. I don't know how viable this fix would be, or whether it is the right fix, but it would prevent the crash in all possible places.
> > 
> > 
> > That sounds like lots of places would need to be updated to do this.
> 
> No, only in one place. The idea I had -- and did a quick prototype of -- was to change the implementation of `[ThreadUtilities getJNIEnv]` to check a flag to see whether or not AWT was the owner of the NSApplication and, if not, effectively do the same thing as `getJNIEnvUncached`. What that would then do would be to re-attach the AppKit thread to the JVM if FX had detached it. This will avoid the crash in all places without having to change the callers of `getJNIEnv`
 
I've looked and I think that would work. Minimal overhead in the 99.99% "non-embedded" case.

Reply:
Here is a Draft PR that fixes this in FX, if we decide to go that route: openjdk/jfx#1545

I added three new regression tests. Two of them (including one that reproduces this crash) fail without the fix. All three pass with the fix.

If you prefer, we can continue to explore whether or not to fix this in AWT in addition to or instead of the AWT fix.

Reply:
> If you prefer, we can continue to explore whether or not to fix this in AWT in addition to or instead of the AWT fix.

typo: I meant "...in addition to or instead of the FX fix."

Reply:
I am starting to think it best to do both of these. It will improve interop of both toolkits with other toolkits, besides each other, and you get the problem fixed no matter whether you have a fixed JDK or a fixed FX.

Reply:
That makes sense to me. I sent a quick (definitely not finished) prototype of the idea I mentioned to @prsadhuk offline, so maybe he can take that as a starting point for the fix.

Reply:
If we do fix it in both, we will need two bug IDs. I can file a new one for JavaFX in that case.

Reply:
NSApplication owner checked for reattaching the thread..CI testing is green...